Replace data vending service with SendTransactionFlow

patrickkuo opened the pull request:
* A new SendTransactionFlow to enable ad hoc data access for resolving transaction.
* Removed Data vending handlers (FetchTransactionsHandler, FetchAttachmentHandler).

shamsasari commented on core/src/main/kotlin/net/corda/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,67 @@
+    @Suspendable
+    override fun call() {
+        // Expect fetch data request until received an end request.
+        while (true) {
```
Why does this loop?

shamsasari commented on core/src/main/kotlin/net/corda/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,67 @@
+    override fun call() {
+        // Expect fetch data request until received an end request.
+        while (true) {
+            val request = getRequest(payload).unwrap {
```
This loop is very hard to follow. Further `getRequest` does more than just "getting a request"

shamsasari commented on core/src/main/kotlin/net/corda/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,67 @@
+    override fun getRequest(payload: Any?) = sendAndReceiveWithRetry<FetchDataFlow.Request>(otherSide, payload!!)
+}
+
+abstract class AbstractSendTransactionFlow(private var payload: Any? = null) : FlowLogic<Unit>() {
```
Use a local variable in `call` and keep `payload` a val

patrickkuo commented:
Thanks @shamsasari, I have address all the issues, back to you.

shamsasari commented on core/src/main/kotlin/net/corda/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,35 @@
+    } ?: receive<FetchDataFlow.Request>(otherSide)
+
+    @Suspendable
+    override fun verifyRequest(request: FetchDataFlow.Request) = require(requestVerifier(request))
```
This protected method is sufficient for developers to use. No need for `requestVerifier` lambda. `SignTransactionFlow` follows the same pattern.

shamsasari commented on core/src/main/kotlin/net/corda/core/internal/dataVending/DataVending.kt:
```diff
@@ -0,0 +1,58 @@
+import net.corda.flows.SendTransactionFlow
+
+/**
+ * The [SendTransactionFlow] is expected to receive [FetchDataFlow.Request] continuously until the [otherSide] has all the data
```
This KDoc isn't needed

shamsasari commented on core/src/main/kotlin/net/corda/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,35 @@
+ * checks to make sure data request is relevant to the flow. Defaulted to
+ */
+
+class SendTransactionFlow(val otherSide: Party,
```
ServeTransactionDataFlow?

mikehearn replied:
Why doesn't this flow take care of sending the transaction as well? In the cases where it's being used it almost always proceeds a message that contains the transaction itself. It's a slightly odd API that you send a TX and then have to remember to invoke a flow. Why not just pass the tx as a parameter and send it as the first step?

shamsasari replied:
This was something I was going to suggest, but realised the other side, ResolveTransactionFlow, should do a receive as a first step. But the current usage of ResolveTransactionFlow might make that difficult.

patrickkuo replied:
Yes, thats what I had initially, but reverted it because it will look different from ResolveTransactionFlow... I will see if there are anything I can do with ResolveTransactionFlow.

adagys commented on docs/source/example-code/src/main/kotlin/net/corda/docs/FxTransactionBuildTutorial.kt:
```diff
@@ -161,7 +161,10 @@ class ForeignExchangeFlow(val tradeId: String,
-        val allPartySignedTx = sendAndReceive<DigitalSignature.WithKey>(remoteRequestWithNotary.owner, signedTransaction).unwrap {
+        send(remoteRequestWithNotary.owner, signedTransaction)
+        // Allow otherParty to access our data to resolve the transaction.
+        subFlow(SendTransactionFlow(remoteRequestWithNotary.owner))
```
maybe we should detect when a SignedTransaction is being sent and auto-invoke the SendTxFlow? I'm thinking of doing something similar on the receiving end

patrickkuo replied:
yes, there is a conversation at our corner about if we should hide `ResolveTransactionFlow`, as its too low level... @mikehearn whats your thought?

mikehearn replied:
I like the idea in theory but it needs a bit of thought - perhaps a mini design sketch in a discourse post or something (not a full blown design doc).

mikehearn replied:
For now we can just replace the send/subFlow dance with a single subFlow call and move the send() into the subflow. Seems more intuitive to me? It is called SendTransactionFlow after all.

mikehearn commented:
This PR will have merge conflicts with PR #1024  -  as the latter may need a backport, please wait for 1024 to go in and rebase on top of it before merging this. Sorry for the pain!

mikehearn commented on core/src/main/kotlin/net/corda/core/internal/dataVending/DataVending.kt:
```diff
@@ -0,0 +1,58 @@
+package net.corda.core.internal.dataVending
```
it's not really standard to have uppercase letters in package names ... partly because domain names aren't case sensitive.

mikehearn commented on core/src/main/kotlin/net/corda/core/internal/dataVending/FetchAttachmentsFlow.kt:
```diff
@@ -38,3 +34,6 @@ class FetchAttachmentsFlow(requests: Set<SecureHash>,
 }
+
+@CordaSerializable
+data class FetchAttachmentsRequest(override val hashes: List<SecureHash>) : FetchDataFlow.Request
```
Perhaps this should be inside FetchAttachmentsFlow, so it can then just be called Request and is namespaced by the enclosing class, like FetchDataFlow.Request is (we should have a convention for this).

mikehearn commented on core/src/main/kotlin/net/corda/core/internal/dataVending/FetchDataFlow.kt:
```diff
@@ -53,55 +56,54 @@ abstract class FetchDataFlow<T : NamedByHash, in W : Any>(
-                fromDisk += stx
-        }
-        return Pair(fromDisk, toFetch)
+        return requests.map { it to load(it) }
```
Although shorter I  am unconvinced this is clearer.

mikehearn commented on core/src/main/kotlin/net/corda/core/internal/dataVending/FetchTransactionsFlow.kt:
```diff
@@ -13,9 +12,9 @@ import net.corda.core.transactions.SignedTransaction
-class FetchTransactionsFlow(requests: Set<SecureHash>, otherSide: Party) :
-        FetchDataFlow<SignedTransaction, SignedTransaction>(requests, otherSide) {
-
+internal class FetchTransactionsFlow(requests: Set<SecureHash>, otherSide: Party, sendEndRequest: Boolean = true) : FetchDataFlow<SignedTransaction, SignedTransaction>(requests, otherSide, sendEndRequest) {
```
I am not sure we should mark flows as internal. The problem is that this will cause the class name to be mangled, which will show up in the wire protocol, and therefore in any eventual spec we publish. Actually I'm pretty sure we shouldn't ... it means if we move a flow between modules, the wire protocol will change.

mikehearn commented on core/src/main/kotlin/net/corda/core/internal/dataVending/FetchDataFlow.kt:
```diff
@@ -53,55 +56,54 @@ abstract class FetchDataFlow<T : NamedByHash, in W : Any>(
     }
 }
+
+class EndDataRequest(override val hashes: List<SecureHash> = emptyList()) : FetchDataFlow.Request
```
Ditto.

patrickkuo commented:
I have made `SendTransactionFlow` send the initial message to avoid the send -> subFlow dance.
I have addressed most of the issues, i think this PR is ready for more review.
Thanks.

mikehearn commented on core/src/main/kotlin/net/corda/flows/BroadcastTransactionFlow.kt:
```diff
@@ -28,7 +28,8 @@ class BroadcastTransactionFlow(val notarisedTransaction: SignedTransaction,
         val msg = NotifyTxRequest(notarisedTransaction)
         participants.filter { it != serviceHub.myInfo.legalIdentity }.forEach { participant ->
             // This pops out the other side in NotifyTransactionHandler
```
Is the comment about "popping out" still correct?

patrickkuo commented:
Added `ResolvableTransactionData` interface and changed SendTransactionFlow and ResolveTransactionsFlow to  `SendTransactionFlow(otherSide: Party, data: ResolvableTransactionData?)` and `ResolveTransactionsFlow(otherSide: Party, data: ResolvableTransactionData)` respectively.

Also added `ReceiveTransactionFlow` which is a combination of `receive` and `ResolveTransactionsFlow`, tis flow lines up with `SendTransactionFlow` nicely.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FetchAttachmentsFlow.kt:
```diff
@@ -13,9 +13,9 @@ import net.corda.core.serialization.SerializeAsTokenContext
  * attachments are saved to local storage automatically.
  */
-@InitiatingFlow
 class FetchAttachmentsFlow(requests: Set<SecureHash>,
```
Can you move  this into the same file as ResolveTransactionsFlow as that it become private

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FetchTransactionsFlow.kt:
```diff
@@ -12,9 +18,8 @@ import net.corda.core.transactions.SignedTransaction
  * the database, because it's up to the caller to actually verify the transactions are valid.
  */
-@InitiatingFlow
 class FetchTransactionsFlow(requests: Set<SecureHash>, otherSide: Party) :
```
Likewise for this as well

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/ResolveTransactionsFlow.kt:
```diff
@@ -152,49 +127,55 @@ class ResolveTransactionsFlow(private val txHashes: Set<SecureHash>,
     }
 }
+
+@CordaSerializable
```
You can remove this annotation from all sub-types

mikehearn replied:
This is new API and thus *must* have API docs explaining what it's for - and you need to update the docsite too! New public API is not to be taken lightly at this stage.

I don't really get the name - what's this for, really? My original proposal was to just separate out at the sendAndReceive level the transaction and the extra stuff ... here it seems we're expecting them to be combined but then implement an interface that separates them out again? Is this really simpler?

patrickkuo replied:
Agree...
I was trying to have 1 flow to handle both SignedTransaction and Hashes, it will be much clearer if I separate them, and remove the need of this interface.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,70 @@
+ */
+open class SendTransactionFlow(protected val otherSide: Party, protected val data: ResolvableTransactionData?) : FlowLogic<Unit>() {
+    @Suspendable
+    protected open fun sendPayloadAndReceiveDataRequest(payload: Any?) = payload?.let { sendAndReceive<FetchDataFlow.Request>(otherSide, payload) } ?: receive<FetchDataFlow.Request>(otherSide)
```
if-else

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,70 @@
+                when (request) {
+                    is FetchDataFlow.Request.Data -> {
+                        // Verify request.
+                        if (request.hashes.isEmpty()) throw FlowException("Empty hash list")
```
Not needed if `hashes` is NonEmptySet

shamsasari commented on docs/source/example-code/src/main/kotlin/net/corda/docs/CustomNotaryTutorial.kt:
```diff
@@ -44,45 +43,25 @@ class MyValidatingNotaryFlow(otherSide: Party, service: MyCustomValidatingNotary
+        } catch(e: Exception) {
             throw when (e) {
-                is TransactionVerificationException,
+                is TransactionVerificationException -> NotaryException(NotaryError.TransactionInvalid(e))
```
Change this into multiple catches

patrickkuo commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,35 @@
+ * which is a sub class of [ResolvableTransactionData]. This flow will resolve the transaction data and return a [UntrustworthyData]
+ * for further verification.
+ */
+class ReceiveTransactionFlow<T : ResolvableTransactionData>(private val expectedTXClazz: Class<T>,
```
ideally should be using `ReceiveTransactionFlow` from now on, I will add a @Deprecated to ResolveTransactionsFlow

patrickkuo commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,35 @@
+}
+
+// Helper method for Kotlin.
+inline fun <reified T : ResolvableTransactionData> FlowLogic<*>.receiveTransaction(otherSide: Party,
```
I am marking these helper methods as internal

patrickkuo commented:
Thanks @shamsasari , all issues addressed.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/BroadcastTransactionFlow.kt:
```diff
@@ -18,16 +18,16 @@ import net.corda.core.utilities.NonEmptySet
-    @CordaSerializable
-    data class NotifyTxRequest(val tx: SignedTransaction)
+
+    data class NotifyTxRequest(override val stx: SignedTransaction) : ResolvableTransactionData.Transaction
```
You don't need this request object anymore do you? I believe SignedTransaction implements your interface?

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FetchDataFlow.kt:
```diff
@@ -30,7 +31,7 @@ import java.util.*
 abstract class FetchDataFlow<T : NamedByHash, W : Any>(
```
Can this be made private as well?

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,39 @@
+                                                            private val verifySignatures: Boolean,
+                                                            private val verifyTransaction: Boolean) : FlowLogic<UntrustworthyData<T>>() {
+
+    constructor(expectedTXClazz: Class<T>, otherSide: Party) : this(expectedTXClazz, otherSide, true)
```
Actually you can do this: `@JvmOverloads constructor(...`

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/ResolveTransactionsFlow.kt:
```diff
@@ -152,49 +138,98 @@ class ResolveTransactionsFlow(private val txHashes: Set<SecureHash>,
+        val stx: SignedTransaction
+    }
+
+    data class TransactionHashes(override val dependencies: Set<SecureHash>) : ResolvableTransactionData
```
This is only used in FlowCookbookJava and some tests. Perhaps just give each of them a copy and remove this?

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,74 @@
+    @Suspendable
+    protected open fun sendPayloadAndReceiveDataRequest(payload: Any?): UntrustworthyData<FetchDataFlow.Request> {
+        return if (payload != null) {
+            payload.let { sendAndReceive<FetchDataFlow.Request>(otherSide, payload) }
```
`let` is no longer needed

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,74 @@
+
+// Convenient methods for Kotlin.
+@Suspendable
+internal fun FlowLogic<*>.sendTransaction(otherSide: Party, data: ResolvableTransactionData?) = subFlow(SendTransactionFlow(otherSide, data))
```
Why is `data` nullable if this is a send tx call?

patrickkuo replied:
in a few test we start the data vending service without sending any transaction, the nullable is to cater that... i will try to modify the test.

shamsasari commented on docs/source/changelog.rst:
```diff
@@ -45,6 +45,9 @@ UNRELEASED
   It is intended that new helper methods will be added shortly to the API to reduce the impact of these changes.
 
+* Data Vending service has been removed in this release to enhance data security. CorDapp developers should use the new
+  ``SendTransactionFlow`` in the flow logic to allow data access from their corresponding counter-flow.
```
Also mention the other side flow

shamsasari commented on docs/source/example-code/src/main/kotlin/net/corda/docs/FlowCookbook.kt:
```diff
@@ -382,12 +382,12 @@ object FlowCookbook {
             // using``ResolveTransactionsFlow`` before verifying it.
             // DOCSTART 13
-            subFlow(ResolveTransactionsFlow(twiceSignedTx, counterparty))
+          //  subFlow(ResolveTransactionsFlow(twiceSignedTx, counterparty))
```
Should there be a TODO to fix this later? You'll need to notify @joeldudleyr3 of these changes anyway.

mikehearn commented on core/src/main/kotlin/net/corda/core/flows/BroadcastTransactionFlow.kt:
```diff
@@ -18,16 +17,12 @@ import net.corda.core.utilities.NonEmptySet
         participants.filter { it != serviceHub.myInfo.legalIdentity }.forEach { participant ->
-            // This pops out the other side in NotifyTransactionHandler
-            send(participant, msg)
+            // SendTransactionFlow allows otherParty to access our data to resolve the transaction.
```
Comment no longer matches the code.

mikehearn commented on core/src/main/kotlin/net/corda/core/flows/CollectSignaturesFlow.kt:
```diff
@@ -125,7 +125,8 @@ class CollectSignaturesFlow(val partiallySignedTx: SignedTransaction,
      */
     @Suspendable private fun collectSignature(counterparty: Party): DigitalSignature.WithKey {
-        return sendAndReceive<DigitalSignature.WithKey>(counterparty, partiallySignedTx).unwrap {
+        // SendTransactionFlow allows otherParty to access our data to resolve the transaction.
```
Comment no longer matches the code.

mikehearn commented on core/src/main/kotlin/net/corda/core/flows/CollectSignaturesFlow.kt:
```diff
@@ -185,14 +186,13 @@ abstract class SignTransactionFlow(val otherParty: Party,
-        val checkedProposal = receive<SignedTransaction>(otherParty).unwrap { proposal ->
+        // Receive transaction and resolve dependencies, check signatures is disabled as we don't have all signatures.
+        val checkedProposal = receiveTransaction<SignedTransaction>(otherParty, verifySignature = false).unwrap { proposal ->
             progressTracker.currentStep = VERIFYING
```
This progress tracker step should now be moved up or removed, as it no longer wraps long-running operations like it did before.

mikehearn commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,39 @@
+
+class ReceiveTransactionFlow<out T : ResolvableTransactionData>
+@JvmOverloads
+constructor(private val expectedTXClazz: Class<T>,
```
This can be called expectedTransactionClass - clazz is only needed when it's a standalone identifier (because it'd conflict with the reserved word)

mikehearn commented on core/src/main/kotlin/net/corda/core/flows/ResolveTransactionsFlow.kt:
```diff
@@ -152,49 +135,52 @@ class ResolveTransactionsFlow(private val txHashes: Set<SecureHash>,
+
         while (nextRequests.isNotEmpty()) {
+            if (limitCounter < 0) throw ExcessivelyLargeTransactionGraph()
+            limitCounter -= nextRequests.size
```
The original formulation used an overflow-checked add - can you use an overflow checked subtract if you want to change this?

mikehearn commented on core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,69 @@
+
+    @Suspendable
+    protected open fun verifyDataRequest(dataRequest: FetchDataFlow.Request.Data) {
+        // TODO: Verify request is relevant to the transaction.
```
Can you expand on this that it's a security related todo? Maybe add the word "security" to it?

mikehearn commented on core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,69 @@
+
+// Convenient methods for Kotlin.
+@Suspendable
+internal fun FlowLogic<*>.sendTransaction(otherSide: Party, data: ResolvableTransactionData) = subFlow(SendTransactionFlow(otherSide, data))
```
I can see why you introduced these, but I'm not sure they're worth it:

1. They're marked as internal only, however, they appear in code that users may look at to learn how flows work. Why are we defining this for ourselves, but then not letting users have it? 
2. They save a very small number of keypresses.
3. They make it less clear what's going on, I think, and they have no docs that a reader can use to quickly Ctrl-J his/her way to enlightenment.
4. Introducing them made a bunch of comments obsolete (see above)

If you just inlined these helpers to all the use sites would we really lose much? At least then there'd be exactly one way to do things which is the same for us as for our users ...

mikehearn commented on core/src/test/kotlin/net/corda/core/flows/AttachmentTests.kt:
```diff
@@ -58,6 +62,8 @@ class AttachmentTests {
         val nodes = mockNet.createSomeNodes(2)
         val n0 = nodes.partyNodes[0]
         val n1 = nodes.partyNodes[1]
+        n0.registerInitiatedFlow(FetchAttachmentsResponse::class.java)
```
Should we just do this automatically for mock nodes?

adagys commented on docs/source/example-code/src/main/java/net/corda/docs/FlowCookbookJava.java:
```diff
@@ -397,12 +399,12 @@ public Void call() throws FlowException {
             // using``ResolveTransactionsFlow`` before verifying it.
             // DOCSTART 13
-            subFlow(new ResolveTransactionsFlow(twiceSignedTx, counterparty));
+            subFlow(new ResolveTransactionsFlow(counterparty, twiceSignedTx));
```
this won't work anymore right? maybe we should remove it from the cookbook

adagys commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,44 @@
+ * [UntrustworthyData] for further verification.
+ */
+
+class ReceiveTransactionFlow<out T : Any>
```
what's the advantage of having this as a flow instead of say a `receiveTransaction()` method on `FlowLogic`?

mikehearn replied:
It's a minor consistency point that's all.

adagys commented on docs/source/example-code/src/main/kotlin/net/corda/docs/CustomNotaryTutorial.kt:
```diff
@@ -44,45 +43,23 @@ class MyValidatingNotaryFlow(otherSide: Party, service: MyCustomValidatingNotary
         try {
-            stx.verifySignaturesExcept(serviceHub.myInfo.notaryIdentity.owningKey)
-        } catch(e: SignatureException) {
+            return subFlow(ReceiveTransactionFlow(SignedTransaction::class.java, otherSide, verifySignatures = false)).unwrap {
```
these callsites seem very verbose, maybe we can add some helpers (that also work in Java) for this?

patrickkuo commented:
I have removed the ResolvableTransactionData interface and changed SendTransactionFlow to `SendTransactionFlow(otherSide:Party,stx:SignedTransaction ,extraData:Any?)`.

There are some casting in ReceiveTransactionFlow to make this work as extraData is nullable. I am not so sure if this is the right way.... Please share your thoughts.

The code is ready for more review. I will update the doc in parallel.

adagys commented on docs/source/example-code/src/main/kotlin/net/corda/docs/FlowCookbook.kt:
```diff
@@ -382,12 +383,12 @@ object FlowCookbook {
             // using``ResolveTransactionsFlow`` before verifying it.
             // DOCSTART 13
-            subFlow(ResolveTransactionsFlow(twiceSignedTx, counterparty))
+            subFlow(ResolveTransactionsFlow(counterparty, twiceSignedTx))
```
same, probably needs to be removed. Also might be worth adding the new send/receive tx flows here

mikehearn commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,44 @@
+ * The [ReceiveTransactionFlow] should be called in response to the [SendTransactionFlow]. It automates the receiving
+ * and resolving of a signed transaction or input hashes.
+ *
+ * This flow is a combination of [receive] and [ResolveTransactionsFlow], it will expect a incoming message of type
```
I don't understand this API anymore :( The docs say it expects TransactionData but the constructor seems to require that you specify explicitly what it expects, and the use sites specify SignedTransaction and other such things.

patrickkuo replied:
TransactionData is used internally to encapsulate the signed transaction/hashes and the extra data, I will rephrase the docs to make it more clear.

patrickkuo replied:
Ok.... after some thinking I think it make more sense to separated out SignedTransaction and extraData, I tried to put both in 1 flow and this makes the API looks confusing, I realised that when I try to rewrite the docs for ReceiveTransactionFlow..... I am planning to change `SendTransactionFlow` to only sending `SignedTransaction` or `Set<SecureHash>`, the extra data will be in a different sub flow.

patrickkuo commented:
I have separated out extra data from SendTransactionFlow to simplify the API, extraData are sent after SendTransactionFlow if necessary.

I have also added a new `SendProposalFlow` to send `TradeProposal` containing StateAndRef, the `StateAndRef` will needed to be resolved like the `SignedTransaction` in `TwoPartyTradeFlow` and FXFlow, but they are completely different use cases, I though it will be more clear if we separate them.

This branch is still WIP, not ready for review yet, will need to tidy up the docs and rebase.

patrickkuo commented:
`SendTransactionFlow` now only send a `SignedTransaction` to the other side and the `ReceiveTransactionFlow` receives it and resolves the dependancies. The user will need to do an extra send if they want to send extra data over (Proposal etc...). 

I think the API for sending and receive/resolving transaction looks more consistent now, this PR is ready for more review.....

Thanks!

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/AbstractStateReplacementFlow.kt:
```diff
@@ -114,9 +114,10 @@ abstract class AbstractStateReplacementFlow {
-            return response.unwrap {
+            val proposal = Proposal(originalState.ref, modification)
+            subFlow(SendTransactionFlow(party, stx))
+            send(party, proposal)
```
this can be sendAndReceive

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/AbstractStateReplacementFlow.kt:
```diff
@@ -162,7 +163,6 @@ abstract class AbstractStateReplacementFlow {
             checkMySignatureRequired(stx)
-            checkDependenciesValid(stx)
             // We expect stx to have insufficient signatures here
             stx.verify(serviceHub, checkSufficientSignatures = false)
```
It makes for ReceiveTransactionFlow to do this, right? And perhaps you pass in the `checkSufficientSignatures` boolean?

mikehearn commented on docs/source/example-code/src/main/kotlin/net/corda/docs/FxTransactionBuildTutorial.kt:
```diff
@@ -227,23 +226,23 @@ class ForeignExchangeRemoteFlow(val source: Party) : FlowLogic<Unit>() {
-        val ourKey = serviceHub.keyManagementService.filterMyKeys(ourResponse.inputs.flatMap { it.state.data.participants }.map { it.owningKey }).single()
-        val proposedTrade = sendAndReceive<SignedTransaction>(source, ourResponse).unwrap {
+        val ourKey = serviceHub.keyManagementService.filterMyKeys(ourInputState.flatMap { it.state.data.participants }.map { it.owningKey }).single()
+        // SendTransactionFlow allows otherParty to access our data to resolve the transaction.
```
Comment uses wrong flow name.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/AbstractStateReplacementFlow.kt:
```diff
@@ -195,7 +195,7 @@ abstract class AbstractStateReplacementFlow {
          * proposal is returned if acceptable, otherwise a [StateReplacementException] is thrown.
```
Update the doc here to include the stx

mikehearn commented on finance/src/main/kotlin/net/corda/flows/TwoPartyTradeFlow.kt:
```diff
@@ -177,23 +170,23 @@ object TwoPartyTradeFlow {
 
         @Suspendable
-        private fun assembleSharedTX(tradeRequest: SellerTradeInfo): Pair<TransactionBuilder, List<PublicKey>> {
+        private fun assembleSharedTX(assetForSale:StateAndRef<OwnableState>, tradeRequest: SellerTradeInfo): Pair<TransactionBuilder, List<PublicKey>> {
```
space after :

mikehearn commented on node/src/integration-test/kotlin/net/corda/node/services/statemachine/LargeTransactionsTest.kt:
```diff
@@ -34,16 +34,16 @@ class LargeTransactionsTest {
             // Send to the other side and wait for it to trigger resolution from us.
-            sendAndReceive<Unit>(serviceHub.networkMapCache.getNodeByLegalName(BOB.name)!!.legalIdentity, stx)
+            subFlow(SendTransactionFlow(serviceHub.networkMapCache.getNodeByLegalName(BOB.name)!!.legalIdentity, stx))
+            receive<Unit>(serviceHub.networkMapCache.getNodeByLegalName(BOB.name)!!.legalIdentity)
```
Maybe pull the node out into a val.

mikehearn commented on test-utils/src/main/kotlin/net/corda/testing/DataVendingFlow.kt:
```diff
@@ -0,0 +1,38 @@
+import net.corda.core.internal.FetchDataFlow
+import net.corda.core.utilities.unwrap
+
+// Flow to start data vending without sending transaction. Copied from [SendTransactionFlow], for testing only.
```
Mmmm .... is copying the code this way really a good idea? Can't we use inheritance instead? Otherwise SendTransactionFlow will get out of sync with this copy and it won't test what we think it tests.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,81 @@
+ * The [SendTransactionFlow] should be called in response to the [ReceiveTransactionFlow]. This flow sends the
+ * [SignedTransaction] to the [otherSide].
+ *
+ * After sending the [SignedTransaction], the [SendTransactionFlow] will listen for incoming [FetchDataFlow.Request]
```
`FetchDataFlow.Request` is invisible to the developer. Just mention that this flow sends back any tx data that's needed by the other side, or words to that effect.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,81 @@
+ * @param otherSide the target party.
+ * @param stx the [SignedTransaction] being sent to the [otherSide].
+ */
+class SendTransactionFlow(otherSide: Party, stx: SignedTransaction) : DataVendingFlow<SignedTransaction>(otherSide, stx)
```
These two flows need to be `open` to be able to override `verifyDataRequest`

shamsasari commented on docs/source/example-code/src/main/java/net/corda/docs/FlowCookbookJava.java:
```diff
@@ -391,18 +396,36 @@ public Void call() throws FlowException {
+            subFlow(new SendTransactionFlow(counterparty, twiceSignedTx) {
+                @Override
+                protected void verifyDataRequest(@NotNull FetchDataFlow.Request.Data dataRequest) {
+                    super.verifyDataRequest(dataRequest);
```
I just a had a look at `super.verifyDataRequest(dataRequest);` and it seems it will have some security. Forgetting to call `super` will bypass that. I think the simple solution is to move the TODO out of verifyDataRequest and remove the call to `super` here

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt:
```diff
@@ -0,0 +1,68 @@
+ */
+open class SendStateAndRefFlow(otherSide: Party, stateAndRefs: List<StateAndRef<*>>) : DataVendingFlow<List<StateAndRef<*>>>(otherSide, stateAndRefs)
+
+sealed class DataVendingFlow<out T : Any>(val otherSide: Party, val payload: T) : FlowLogic<Unit>() {
```
The type parameter `T` isn't actually used

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,47 @@
+ * and perform the resolution back-and-forth required to check the dependencies.
+ * The flow will return the list of [StateAndRef] after it is resolved.
+ *
+ * @JvmSuppressWildcards is used to suppress wildcards in return type when calling `subFlow(new ReceiveStateAndRef<T>(otherParty))` in java.
```
This is not documentation, but just a `//` comment (what do you call that comment type??)

shamsasari replied:
This is not documentation, but just a `//` comment (what do you call that comment type??)

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,47 @@
+@JvmOverloads
+constructor(private val otherParty: Party, private val checkSufficientSignatures: Boolean = true) : FlowLogic<SignedTransaction>() {
+    @Suspendable
+    override fun call(): SignedTransaction {
```
You need to add the same `@Throws` annotation from `SignedTransaction.verify`, for our Java friends.

shamsasari commented on node/src/main/kotlin/net/corda/node/services/CoreFlowHandlers.kt:
```diff
@@ -5,69 +5,23 @@ import net.corda.core.contracts.ContractState
-        request.tx.verify(serviceHub)
-        serviceHub.recordTransactions(request.tx)
+        val stx = subFlow(ReceiveTransactionFlow(otherParty))
+        stx.verify(serviceHub)
```
This isn't needed

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,47 @@
+ *
+ * This flow is a combination of [receive] and [ResolveTransactionsFlow]. This flow will receive the [SignedTransaction]
+ * and perform the resolution back-and-forth required to check the dependencies and download any missing attachments.
+ * The flow will return the [SignedTransaction] after it is resolved and verified.
```
after it is resolved and then verified using [SignedTransaction.verify].

patrickkuo commented:
Thanks @shamsasari, I have did a rebase and addressed the issues.

shamsasari commented on docs/source/api-flows.rst:
```diff
@@ -478,11 +479,29 @@ transaction and provide their signature if they are satisfied:
-using``ResolveTransactionsFlow`` before verifying it:
+SendTransactionFlow/ReceiveTransactionFlow
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Verifying a transaction will also verify every transaction in the transaction's dependency chain, which will require
```
Verifying a transaction received from a counterparty also requires verification of every transaction in its dependency chain. This means the receiving party needs to be able to ask the sender all the details of the chain. The sender will use ``SendTransactionFlow`` for sending the transaction and then for processing all subsequent transaction data vending requests as the receiver walks the dependency chain using ``ReceiveTransactionFlow``.

shamsasari commented on docs/source/api-flows.rst:
```diff
@@ -498,7 +517,7 @@ using``ResolveTransactionsFlow`` before verifying it:
         :dedent: 12
 
-We can also resolve a `StateRef` dependency chain:
+We can also send and receive a `StateAndRef` dependency chain and automatically resolve its dependencies:
```
double back tick surrounding StateAndRef

shamsasari commented on docs/source/example-code/src/main/kotlin/net/corda/docs/FlowCookbook.kt:
```diff
@@ -375,18 +376,35 @@ object FlowCookbook {
+            // Optional request verification to further restrict data access.
+            subFlow(object :SendTransactionFlow(counterparty, twiceSignedTx){
+                override fun verifyDataRequest(dataRequest: FetchDataFlow.Request.Data) {
+                    super.verifyDataRequest(dataRequest)
```
You can remove this

shamsasari commented on core/src/test/kotlin/net/corda/core/flows/ResolveTransactionsFlowTest.kt:
```diff
@@ -1,14 +1,14 @@
 import net.corda.core.crypto.SecureHash
 import net.corda.core.getOrThrow
 import net.corda.core.identity.Party
+import net.corda.core.internal.ResolveTransactionsFlow
```
You need to move this test class as well.

shamsasari commented on docs/source/changelog.rst:
```diff
@@ -72,6 +72,14 @@ UNRELEASED
     * ``stateMachineRecordedTransactionMapping``, replaced by ``stateMachineRecordedTransactionMappingFeed``
     * ``networkMapUpdates``, replaced by ``networkMapFeed``
 
+* Data Vending service has been removed in this release to enhance data security, ``FetchTransactionFlow`` and
```
How about:
Due to security concerns and the need to remove the concept of state relevancy (which isn't needed in Corda), ``ResolveTransactionsFlow`` has been made internal. Instead merge the receipt of the ``SignedTransaction`` and the subsequent sub-flow call to ``ResolveTransactionsFlow`` with a single call to ``ReceiveTransactionFlow``. The flow running on the counterparty must use ``SendTransactionFlow`` at the correct place. There is also ``ReceiveStateAndRefFlow`` and ``SendStateAndRefFlow`` for dealing with ``StateAndRef``s.

shamsasari commented on test-utils/src/main/kotlin/net/corda/testing/DataVendingFlow.kt:
```diff
@@ -0,0 +1,20 @@
+import net.corda.core.utilities.UntrustworthyData
+
+// Flow to start data vending without sending transaction. For testing only.
+class DataVendingFlow(otherSide: Party) : SendStateAndRefFlow(otherSide, emptyList()) {
```
Can we move this to core/test as it's only used there.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,48 @@
+/**
+ * The [ReceiveTransactionFlow] should be called in response to the [SendTransactionFlow].
+ *
+ * This flow is a combination of [receive] and [ResolveTransactionsFlow]. This flow will receive the [SignedTransaction]
```
ResolveTransactionsFlow is now internal and so shouldn't be referenced.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/ReceiveTransactionFlow.kt:
```diff
@@ -0,0 +1,48 @@
+/**
+ * The [ReceiveStateAndRefFlow] should be called in response to the [SendStateAndRefFlow].
+ *
+ * This flow is a combination of [receive] and [ResolveTransactionsFlow]. This flow will receive a list of [StateAndRef]
```
ditto

shamsasari commented on docs/source/tutorial-attachments.rst:
```diff
@@ -39,9 +39,9 @@ a JVM client.
+Normally attachments on transactions are fetched automatically via the ``ReceiveTransactionFlow``. Attachments
 are needed in order to validate a transaction (they include, for example, the contract code), so must be fetched
-before the validation process can run. ``ResolveTransactionsFlow`` calls ``FetchTransactionsFlow`` to perform the
+before the validation process can run. ``ReceiveTransactionFlow`` calls ``ResolveTransactionsFlow`` to perform the
```
ditto

patrickkuo commented:
@mikehearn do you have more comments?

mikehearn commented:
My comments last time were all minor. If Shams is happy I'm happy!
